subsys/bluetooth/mesh/Kconfig

```diff
@@ -1040,6 +1040,30 @@ config BT_MESH_BLOB_IO_FLASH
 	  Enable the BLOB flash stream for reading and writing BLOBs directly to
 	  and from flash.
 
+if BT_MESH_BLOB_IO_FLASH
+
+config BT_MESH_BLOB_IO_FLASH_WITHOUT_ERASE
+	bool "BLOB flash support for devices without erase"
+	default y if FLASH_HAS_NO_EXPLICIT_ERASE
+	depends on FLASH_HAS_NO_EXPLICIT_ERASE
+	help
+	  Enable path supporting devices without erase. This option appears only
+	  if there are devices without explicit erase requirements in the system
+	  and may be disabled to reduce code size in case when no operations
+	  are intended on such type of devices.
+
+config BT_MESH_BLOB_IO_FLASH_WITH_ERASE
+	bool "BLOB flash support for devices with erase"
+	default y if FLASH_HAS_EXPLICIT_ERASE
+	depends on FLASH_HAS_EXPLICIT_ERASE
+	help
+	  Enable path supporting devices with erase. This option appears only
+	  if there are devices requiring erase,  before write, in the system
+	  and may be disabled to reduce code size in case when no operations
+	  are intended on such type of devices.
```

Comment:
Names of these options are confusing. They and their description look mutually exclusive while they aren't. Can we name them more clearly? For example:
`BT_MESH_BLOB_IO_FLASH_WITH_ERASE` -> `BT_MESH_BLOB_IO_FLASH_SUPPORT_ERASE`
`BT_MESH_BLOB_IO_FLASH_WITHOUT_ERASE` -> `BT_MESH_BLOB_IO_FLASH_SUPPORT_DIRECT_WRITE`.

Also the term `devices` here is a bit confusing as well. Can we specify what exactly we mean here, e.g. `flash memory devices`?

Reply:
Can you not just infer the decision making logic based on the platform's supported memory KCONFIG options? something like those -  SOC_FLASH_NRF_MRAM, CONFIG_SOC_FLASH_NRF_RRAM, etc. 

That we avoid having seperate Kconfig in mesh.

Reply:
> Names of these options are confusing. They and their description look mutually exclusive while they aren't. Can we name them more clearly? For example: `BT_MESH_BLOB_IO_FLASH_WITH_ERASE` -> `BT_MESH_BLOB_IO_FLASH_SUPPORT_ERASE` `BT_MESH_BLOB_IO_FLASH_WITHOUT_ERASE` -> `BT_MESH_BLOB_IO_FLASH_SUPPORT_DIRECT_WRITE`.
> 

No on "Direct write" because there are literally devices that have buffered and direct write operations, and they differ from possibility to write without erase. Also in case of your thing here "with erase" gets additional meaning if you have a device that supports, but does not require erase, as erase may be used, on such devices, to increase write speed of large chunk of data.

> Also the term `devices` here is a bit confusing as well. Can we specify what exactly we mean here, e.g. `flash memory devices`?

Because the entire thing is about non-flash devices and flash devices. They are under one API, but they do not come from the same technology. I still use _FLASH_ in names, but as reference to to API not to device itself.

Reply:
Still, can we make these options more clear for users with less background in this?

Reply:
Sure, I am fine with that. but how?

People without background on what they are doing should not probably tinker with this and leave this default.
The more we do system for "people without background" the less useful as embedded it becomes. The tuning requires understanding what you are doing with hardware you have and understanding that hardware; no layers or Kconfigs can fix lack of understanding.

Reply:
> If this is just an code optimization, can't we rely on the existing options?

I would not call that code optimizations, rather removal of unused paths by lack of support by hardware or support not needed by user

Nevertheless, and what is coming here is my opinion, Zephyr has a huge problem of lasagna code, on the usage of Kconfig options,  where modules cross boundaries of other modules/subsystems to pick information on what is supported and what not. The Flash Kconfigs should not be directly used in code that is not Flash API related that  are supposed to affect the Flash API; these Kconfigs should only be used by Kconfig, as a configuration module that is exposed to them, to impact configuration options of other modules.
This way any changes to module, like Flash, that would mean removal of Kconfigs, would only impact Kconfig options rather than half of Zephyr code.

Whether something is supported by Flash API, still used as an example here, does not mean that subsystem that somehow uses the API may need that, because that subsystem may only invoke, depending on user preferred usage, paths that never rely on some feature of a device/API, yet all the dead code will be included.